src/Microsoft.Diagnostics.Runtime/DataTarget.cs

```diff
@@ -195,15 +195,40 @@ private ImmutableArray<ClrInfo> GetOrCreateClrVersions()
 
             if (_clrs.IsDefault)
             {
+                IEnumerable<ModuleInfo> modules = EnumerateModules();
+                IEnumerable<ClrInfo>? clrs = null;
+
+                // First try the SpecialDiagInfo block short cut to find a supported runtime
+                if (!_target.ForceCompleteRuntimeEnumeration && DataReader.TargetPlatform != OSPlatform.Windows)
+                {
+                    if (SpecialDiagInfo.TryReadSpecialDiagInfo(DataReader, out SpecialDiagInfo info) && info.RuntimeBaseAddress != 0)
+                    {
+                        Debug.Assert(_modules != null);
+                        
+                        // The above EnumerateModules() call populated (the sorted) _modules array
+                        int found = _modules!.Search(info.RuntimeBaseAddress, (m, k) => m.ImageBase.CompareTo(k));
+                        if (found > 0)
+                        {
+                            // Run the module through the clr info providers to see if it is an app model/runtime supported by CLRMD
+                            clrs = from clrInfo in s_clrInfoProviders.Select(provider => provider.ProvideClrInfoForModule(this, _modules![found])).Where(clrInfo => clrInfo != null)
+                                   where clrInfo != null
+                                   orderby clrInfo.Flavor descending, clrInfo.Version
+                                   select clrInfo;
+                        }
+                    }
+                }
+
+                // If no runtimes where found from the SpecialDiagInfo block, run all the modules through the clr info providers
+
                 // We order this so .Net Core comes first, so if there's multiple CLRs we prefer
                 // to debug .Net Core (assuming the user is just debugging one of them)
 
-                IEnumerable<ClrInfo> clrs = from module in EnumerateModules()
-                                            let clrInfoEnumerable = s_clrInfoProviders.Select(provider => provider.ProvideClrInfoForModule(this, module)).Where(clrInfo => clrInfo != null)
-                                            let clrInfo = clrInfoEnumerable.LastOrDefault()
-                                            where clrInfo != null
-                                            orderby clrInfo.Flavor descending, clrInfo.Version
-                                            select clrInfo;
+                clrs ??= from module in modules
```

Comment:
Does desktop CLR have the special diagnostic info block?  It doesn't, right?

That means that if one coreclr is in the process with a special diagnostic block, we will skip enumerating all modules which will find the desktop runtime?  Or am I missing something?

Reply:
No, only Linux and MacOS .NET Core have this block.  The idea is to only return the "crashing" runtime.  If a user wants all runtimes, then the new force flag should be set.  Under SOS this happens in the !runtimes command.

Reply:
Only Linux and MacOS .NET core dumps created with createdump have this block. Windows, live-debugging don't.

Reply:
This isn't quite what I had in mind.  There are lots of scenarios where we don't have a "crashing" runtime.  The change as it stands leads to inconsistent behavior in many scenarios.

I think the right approach is:

1.  Look only for coreclr/clr in the process (under the correct name for all architectures).  That scan is fast, so there's no reason to not look for those runtimes.
2. Disable the SingleFileClrInfoProvider unless the user sets "Force scanning all runtimes."
3. When we have the extra debug info block, use that to supplement 1 and 2.

This means we would always find crashing single-file apps.  We would not find non-crashing single-file apps unless the user requested to scan everything.  More importantly, we will _always_ normal "coreclr" and "clr" runtimes in the process, preserving that behavior.

I am out of office for a while at a team event today.  Let me take a closer look when I'm back tomorrow morning and maybe submit changes to your branch.

Reply:
For all the scenarios that don't have a "crashing" runtime, it is exactly like it was before. It doesn't make sense to me to enumerate any modules when the special diag block exists. Yes, it does change the behavior but the force flag restore the behavior.

I think we need to talk about this some more before you make any changes.